server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -2202,12 +2202,18 @@ protected void flushHoldingLock(boolean force, boolean waitIfOngoing, ActionList
             // newly created commit points to a different translog generation (can free translog),
             // or (4) the local checkpoint information in the last commit is stale, which slows down future recoveries.
             boolean hasUncommittedChanges = hasUncommittedChanges();
+
+            final boolean shouldFlushForPreviousGeneration = store.getFlushByRefreshGenerations()
```

Comment:
Looks like we are trying to address the case where `force` is `false` but the latest commit is not uploaded yet, so we want to flush and wait.

I wonder if we could override `flushHoldingLock(()` in the engine implementation, see if the last committed generation has been uploaded and if not, force the upload by calling the super flushHoldingLock() with force=true?

Reply:
This is a great suggestion. It prompted me to think more how logic can be mostly supplied from the stateless side. I have now made the changes to the stateful side very minimal. Thanks a lot!